### PR TITLE
Fix assorted breakages and Windows compilation changes

### LINKED ISSUE #8351: Externals boringssl target fails to build on Windows
Building on windows, the boringssl library dependency is broken, because it attempts to build both a static and dynamic lib. The dynamic lib built is not desired, and actually not functional. The error is;

ERROR: C:/_eb/external/boringssl/BUILD:130:1: output 'external/boringssl/crypto.if.lib' was not created
ERROR: C:/_eb/external/boringssl/BUILD:130:1: not all outputs were created or valid
ERROR: C:/_eb/external/envoy/bazel/foreign_cc/BUILD:67:1 not all outputs were created or valid

In bazel-bin/external/boringssl/, the build generates;

_objs
crypto.dll
crypto.dll-2.params
crypto.lib
crypto.lib-2.params
crypto.pdb
ssl.dll-2.params
ssl.lib
ssl.lib-2.params

The resulting crypto.dll has no export symbols, all of the boringssl externs are used only for static binding (in linux terms, these are non-pic objects.) As documented on msdn about link.exe, the /implib:crypto.if.lib will never be created when there are no exported symbols. The resulting (desired) crypto.lib is the complete, static library. ssl.dll cannot be created because it requires the missing crypto.if.lib to bind to crypto.dll, which was an invalid shared object library.

We should not be attempting to build dynamic libraries, emit .dll files or their corresponding .if.lib export binding libs, but it is not clear how to suppress this either within envoy's bazel/ logic or within boringssl itself.

@irengrig or @lizan would you have any clues?

### PR DESCRIPTION
**Pull request by sunjayBhatia**
Description:
- Fixes assorted master breakages and changes required for Windows compilation.
- Bumps rapidjson to support Windows and clang-cl
- Moves existing os_sys_call_impl for future parallel posix and win32 implementations
- patch boringssl build to use static library

Risk Level: Low
Testing: Local on gcc Linux, msvc Windows
Docs Changes: N/A
Release Notes: N/A
Fixes #8351

### CODE REVIEW

File: source/common/api/BUILD [hunk 1]
```diff
@@ -22,22 +22,27 @@ envoy_cc_library(
...
+    }) + envoy_select_hot_restart(["posix/os_sys_calls_impl_hot_restart.h"]),
+    strip_include_prefix = select({
+        "//conditions:default": "posix",
+    }),
     deps = [
         "//include/envoy/api:os_sys_calls_interface",
         "//source/common/singleton:threadsafe_singleton",
-    ],
+    ] + select({
+        "//conditions:default": [],
```
Comment (mattklein123) [inline source/common/api/BUILD]: This can be deleted, right?

Comment (wrowe) [reply source/common/api/BUILD]: We'd prefer to leave it (fill-in-the-gap adds windows specifics there in a upcoming patch, if that's alright.) Trying to keep as many changes in the upcoming commits to be line-oriented diffs.

Comment (mattklein123) [reply source/common/api/BUILD]: OK sounds good.

Comment (sunjayBhatia) [reply source/common/api/BUILD]: we found that we didn't need the subsequent changes in the later commit for Windows, so we will remove this bit

Comment (wrowe) [reply source/common/api/BUILD]: We just re-reviewed and our modification turned out to be unnecessary. Follow-up commit removes this select() as requested (and cleans up win32 deps from the follow-on patch).

**Pull request merged**